http/handler.go

```diff
@@ -227,33 +211,43 @@ func respondError(w http.ResponseWriter, status int, err error) {
 	enc.Encode(resp)
 }
 
-func respondCommon(w http.ResponseWriter, resp *logical.Response, err error) bool {
-	if resp == nil {
+func respondErrorCommon(w http.ResponseWriter, resp *logical.Response, err error) bool {
+	// If there are no errors return
+	if err == nil && (resp == nil || !resp.IsError()) {
 		return false
 	}
 
-	if resp.IsError() {
-		statusCode := http.StatusBadRequest
+	// Start out with internal server error since in most of these cases there
+	// won't be a response so this won't be overridden
+	statusCode := http.StatusInternalServerError
+	// If we actually have a response, start out with bad request
+	if resp != nil {
+		statusCode = http.StatusBadRequest
+	}
 
-		if err != nil {
-			switch err {
-			case logical.ErrPermissionDenied:
-				statusCode = http.StatusForbidden
-			case logical.ErrUnsupportedOperation:
-				statusCode = http.StatusMethodNotAllowed
-			case logical.ErrUnsupportedPath:
-				statusCode = http.StatusNotFound
-			case logical.ErrInvalidRequest:
-				statusCode = http.StatusBadRequest
-			}
+	// Now, check the error itself; if it has a specific logical error, set the
+	// appropriate code
+	if err != nil {
+		switch {
+		case errwrap.ContainsType(err, new(vault.StatusBadRequest)):
+			statusCode = http.StatusBadRequest
+		case errwrap.Contains(err, logical.ErrPermissionDenied.Error()):
+			statusCode = http.StatusForbidden
+		case errwrap.Contains(err, logical.ErrUnsupportedOperation.Error()):
+			statusCode = http.StatusMethodNotAllowed
+		case errwrap.Contains(err, logical.ErrUnsupportedPath.Error()):
+			statusCode = http.StatusNotFound
+		case errwrap.Contains(err, logical.ErrInvalidRequest.Error()):
+			statusCode = http.StatusBadRequest
 		}
+	}
 
-		err := fmt.Errorf("%s", resp.Data["error"].(string))
-		respondError(w, statusCode, err)
-		return true
+	if resp != nil {
+		err = fmt.Errorf("%s", resp.Data["error"].(string))
```

Comment:
That's the *current* behavior. I'm not sure it's correct. When returning an error from backends it's totally valid to return both a logical error and an error response and some backends do. Overall the handling isn't the greatest...